### PR TITLE
Add Datastructure unit tests

### PR DESCRIPTION
**Pull request by Alexander-Hjelm**

### CODE REVIEW

File: src/WorkItemMigrator/tests/Migration.Common.Tests/JsonExtensionsTests.cs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        {
+            //Assign
+
+            var fileSystem = _fixture.Freeze<IFileSystem>();
+            fileSystem.File.Exists(Arg.Any<string>()).Returns(true);
+
+            var expected = new Dictionary<string, string>();
+
+            //Act
+            //var actualResult = UserMapper.ParseUserMappings(Arg.Any<string>());
```
Comment (Alexander-Hjelm) [inline src/WorkItemMigrator/tests/Migration.Common.Tests/JsonExtensionsTests.cs:64]: Remove commented-out code

File: src/WorkItemMigrator/tests/Migration.Common.Tests/MigrationCommonTests.cs [hunk 2]
```diff
@@ -49,7 +49,8 @@ public void When_calling_ParseUserMappings_with_non_exisiting_file_Return_empty_
             var expected = new Dictionary<string, string>();
 
             //Act
-            var actualResult = UserMapper.ParseUserMappings(Arg.Any<string>());
+            //var actualResult = UserMapper.ParseUserMappings(Arg.Any<string>());
```
Comment (Alexander-Hjelm) [inline src/WorkItemMigrator/tests/Migration.Common.Tests/MigrationCommonTests.cs:52]: Remove commented-out code

File: src/WorkItemMigrator/tests/Migration.Common.Tests/RevisionUtilityTests.cs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+            DateTime actual = RevisionUtility.NextValidDeltaRev(datetime1, datetime2);
+
+            Assert.AreEqual(expected, actual);
+        }
+
+        [Test]
+        public void When_calling_replacehtmlelements_Then_the_expected_result_is_returned()
+        {
+            string expected = "html";
+            string actual = RevisionUtility.ReplaceHtmlElements("html");
```
Comment (Alexander-Hjelm) [inline src/WorkItemMigrator/tests/Migration.Common.Tests/RevisionUtilityTests.cs:51]: Test two strings that comform to the tested patterns instead, not just "html"
See:
  string imageWrapPattern = "<span class=\"image-wrap\".*?>.*?().*?</span>";
  string userLinkPattern = "<a href=.*? class=\"user-hover\" .*?>(.*?)</a>";

File: src/WorkItemMigrator/tests/Migration.Jira-Export.Tests/JiraLinkTests.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+using JiraExport;
+using AutoFixture.AutoNSubstitute;
+using AutoFixture;
+
+namespace Migration.Jira_Export.Tests
+{
+    [TestFixture]
+    public class JiraLinkTests
+    {
+        // use auto fixiture to help mock and instantiate with dummy data with nsubsitute.
```
Comment (Alexander-Hjelm) [inline src/WorkItemMigrator/tests/Migration.Jira-Export.Tests/JiraLinkTests.cs:12]: do a search replace: "auto fixiture" -> "auto fixture" in the whole project :)

File: src/WorkItemMigrator/tests/Migration.Common.Tests/RevisionUtilityTests.cs (hunk 3 above)
Comment (SusannaP2018) [reply src/WorkItemMigrator/tests/Migration.Common.Tests/RevisionUtilityTests.cs:51]: What would be a good string value to test? I'm having a hard time understandning what Regex.Replace does.

File: src/WorkItemMigrator/tests/Migration.Common.Tests/MigrationCommonTests.cs (hunk 2 above)
Comment (SusannaP2018) [reply src/WorkItemMigrator/tests/Migration.Common.Tests/MigrationCommonTests.cs]: Fixed

File: src/WorkItemMigrator/tests/Migration.Jira-Export.Tests/JiraLinkTests.cs (hunk 4 above)
Comment (SusannaP2018) [reply src/WorkItemMigrator/tests/Migration.Jira-Export.Tests/JiraLinkTests.cs]: Fixed

File: src/WorkItemMigrator/tests/Migration.Common.Tests/RevisionUtilityTests.cs (hunk 3 above)
Comment (Alexander-Hjelm) [reply src/WorkItemMigrator/tests/Migration.Common.Tests/RevisionUtilityTests.cs:51]: You can use https://regex101.com/ to verify the regex expression against a test string.

Here's two string that would work, or smth similar. Test both cases :)

- `<a href=https://text.com class="user-hover" >placeholder string</a>`
- `<span class=\"image-wrap\">()</span>`

Comment (SusannaP2018) [reply src/WorkItemMigrator/tests/Migration.Common.Tests/RevisionUtilityTests.cs]: Added two tests, please take a look when you have time.

File: src/WorkItemMigrator/tests/Migration.Jira-Export.Tests/JiraLinkTests.cs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+        [Test]
+        public void When_calling_to_string_Then_the_expected_string_value_is_returned()
+        {
+            JiraLink sut = new JiraLink();
+            sut.LinkType = "System.LinkTypes.Hierarchy-Forward";
+            sut.SourceItem = "sourceItem";
+            sut.TargetItem = "targetItem";
+
+            string expectedToString = "[" + sut.LinkType + "] " + sut.SourceItem + "->" + sut.TargetItem;
```
Comment (madkoo) [inline src/WorkItemMigrator/tests/Migration.Jira-Export.Tests/JiraLinkTests.cs:30]: Can be writtenn more simpler

File: src/WorkItemMigrator/tests/Migration.Jira-Export.Tests/JiraMapperTests.cs [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+            WiItem expectedWiItem = new WiItem();
+            expectedWiItem.Type = "User Story";
+            expectedWiItem.OriginId = "issue_key";
+
+            JiraMapper sut = createJiraMapper();
+
+            WiItem expected = expectedWiItem;
+            WiItem actual = sut.Map(jiraItem);
+
+            Assert.AreEqual(expected.OriginId, actual.OriginId);
```
Comment (madkoo) [inline src/WorkItemMigrator/tests/Migration.Jira-Export.Tests/JiraMapperTests.cs:43]: wrap multiple asserts into Assert.Multiple(() => {
// your asserts
});

File: src/WorkItemMigrator/tests/Migration.WIContract.Tests/WiFieldTests.cs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+        [Test]
+        public void When_calling_tostring_Then_the_expected_string_value_is_returned()
+        {
+            WiField sut = new WiField();
+
+            sut.ReferenceName = "referenceName";
+            sut.Value = "objValue";
+
+            string expectedToString = "[" + sut.ReferenceName + "]=" + sut.Value;
```
Comment (madkoo) [inline src/WorkItemMigrator/tests/Migration.WIContract.Tests/WiFieldTests.cs:29]: Can be written more simpler

File: src/WorkItemMigrator/tests/Migration.WIContract.Tests/WiItemTests.cs [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+        [Test]
+        public void When_calling_ToString_Then_the_expected_String_value_is_returned()
+        {
+            WiItem sut = new WiItem();
+
+            sut.Type = "type";
+            sut.OriginId = "originId";
+            sut.WiId = 1;
+
+            string expectedToString = "[" + sut.Type + "]"+ sut.OriginId +"/" + sut.WiId;
```
Comment (madkoo) [inline src/WorkItemMigrator/tests/Migration.WIContract.Tests/WiItemTests.cs:30]: Can be written more simpler

File: src/WorkItemMigrator/tests/Migration.WIContract.Tests/WiRevisionTests.cs [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+        [Test]
+        public void When_calling_ToString_Then_the_expected_String_value_is_returned()
+        {
+            WiRevision sut = new WiRevision();
+
+            sut.ParentOriginId = "parentOriginId";
+            sut.Index = 1;
+
+            string expectedToString = "'" + sut.ParentOriginId + "', rev " + sut.Index;
```
Comment (madkoo) [inline src/WorkItemMigrator/tests/Migration.WIContract.Tests/WiRevisionTests.cs:30]: Can be written more simpler
$"'{sut.ParentOriginId}', rev {sut.Index}"

**Review by madkoo — COMMENTED** [review]
Added few comments, please take a look at let me know what do you think or if something is confusing

File: src/WorkItemMigrator/JiraExport/JiraMapper.cs [hunk 10]
```diff
@@ -27,67 +29,8 @@ public JiraMapper(JiraProvider jiraProvider, ConfigJson config) : base(jiraProvi
...
-                else
-                {
-                    Logger.Log(LogLevel.Error, $"Type mapping missing for '{issue.Key}' with Jira type '{issue.Type}'. Item was not exported which may cause missing links in issues referencing this item.");
-                    return null;
-                }
-            }
-            return wiItem;
-        }
-
         public List<WiLink> MapLinks(JiraRevision r)
```
[the remark this answers is not in the document]
Comment (SusannaP2018) [reply src/WorkItemMigrator/JiraExport/JiraMapper.cs]: Kept Map() as internal, changed other methods to internal.

File: src/WorkItemMigrator/WorkItemMigrator.sln [hunk 11]
```diff
@@ -19,7 +19,9 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Migration.Jira-Export.Tests
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Migration.Common.Tests", "tests\Migration.Common.Tests\Migration.Common.Tests.csproj", "{14911A7A-198D-43AD-8E93-D79A3DBBB15E}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Migration.Wi-Import.Tests", "tests\Migration.Wi-Import.Testss\Migration.Wi-Import.Tests.csproj", "{C6099BD7-3590-4E7E-9A7B-5889E33532D3}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Migration.Wi-Import.Tests", "tests\Migration.Wi-Import.Testss\Migration.Wi-Import.Tests.csproj", "{C6099BD7-3590-4E7E-9A7B-5889E33532D3}"
```
[the remark this answers is not in the document]
Comment (SusannaP2018) [reply src/WorkItemMigrator/WorkItemMigrator.sln]: Thanks, fixed :)

File: src/WorkItemMigrator/tests/Migration.Common.Tests/MigrationCommonTests.cs [hunk 12]
```diff
@@ -49,7 +49,7 @@ public void When_calling_ParseUserMappings_with_non_exisiting_file_Return_empty_
             var expected = new Dictionary<string, string>();
 
             //Act
-            var actualResult = UserMapper.ParseUserMappings(Arg.Any<string>());
+            var actualResult = UserMapper.ParseUserMappings("");
```
[the remark this answers is not in the document]
Comment (SusannaP2018) [reply src/WorkItemMigrator/tests/Migration.Common.Tests/MigrationCommonTests.cs:52]: I had an error when running the whole test suite with Arg.Any present in two places:

NSubstitute.Exceptions.RedundantArgumentMatcherException : Some argument specifications (e.g. Arg.Is, Arg.Any) were left over after the last call.

Complete error message: https://pastebin.com/JRnKeaxp

I couldn't figure out how to fix it according to the error message instructions.

I have changed it to string.Empty instead 👍

File: src/WorkItemMigrator/tests/Migration.Jira-Export.Tests/JiraAttachmentTest.cs [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+
+        [Test]
+        public void When_calling_to_string_Then_the_expected_string_value_is_returned()
+        {
+            JiraAttachment sut = new JiraAttachment();
+
+            sut.Id = "id";
+            sut.Filename = "name";
+
+            string expectedToString = sut.Id + "/" + sut.Filename;
```
[the remark this answers is not in the document]
Comment (SusannaP2018) [reply src/WorkItemMigrator/tests/Migration.Jira-Export.Tests/JiraAttachmentTest.cs]: Fixed

File: src/WorkItemMigrator/tests/Migration.Jira-Export.Tests/JiraLinkTests.cs (hunk 5 above)
Comment (SusannaP2018) [reply src/WorkItemMigrator/tests/Migration.Jira-Export.Tests/JiraLinkTests.cs:30]: Fixed

File: src/WorkItemMigrator/tests/Migration.WIContract.Tests/WiFieldTests.cs (hunk 7 above)
Comment (SusannaP2018) [reply src/WorkItemMigrator/tests/Migration.WIContract.Tests/WiFieldTests.cs:29]: Fixed

File: src/WorkItemMigrator/tests/Migration.WIContract.Tests/WiRevisionTests.cs (hunk 9 above)
Comment (SusannaP2018) [reply src/WorkItemMigrator/tests/Migration.WIContract.Tests/WiRevisionTests.cs:30]: Fixed

File: src/WorkItemMigrator/tests/Migration.Jira-Export.Tests/JiraMapperTests.cs (hunk 6 above)
Comment (SusannaP2018) [reply src/WorkItemMigrator/tests/Migration.Jira-Export.Tests/JiraMapperTests.cs:43]: Thanks, fixed :)

File: src/WorkItemMigrator/tests/Migration.Common.Tests/MigrationCommonTests.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (Alexander-Hjelm) [reply src/WorkItemMigrator/tests/Migration.Common.Tests/MigrationCommonTests.cs]: I think string.Empty is a good option :)

[the remark this answers is not in the document]
Comment (madkoo) [reply src/WorkItemMigrator/tests/Migration.Common.Tests/MigrationCommonTests.cs]: Looks good with string.empty

**Review by madkoo — COMMENTED** [review]
Few more notes, please take a look. 
Also when writing Unit tests consider the negative test as well where you input null values or missing values and what happens when code would break.

File: src/WorkItemMigrator/tests/Migration.Common.Tests/JsonExtensionsTests.cs [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+        {
+            JObject jObject = JObject.Parse(@"{ name: 'My Name', emails: [ 'my@email.com', 'my2@email.com' ]}");
+
+            var expected = jObject.SelectToken("emails", false);
+            var actual = JsonExtensions.GetValues<JToken>(jObject, "emails");
+
+            Assert.AreEqual(expected, actual);
+        }
+
+            [Test]
```
[the remark this answers is not in the document]
Comment (Alexander-Hjelm) [reply src/WorkItemMigrator/tests/Migration.Common.Tests/JsonExtensionsTests.cs]: Fixed by 1c998b0

File: src/WorkItemMigrator/tests/Migration.Common.Tests/JsonExtensionsTests.cs [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+            string[] userMapLines = { "a@jira.com=a@azdo.com", "b@jira.com=b@azdo.com" };
+            Dictionary<string, string> generatedUserMap = UserMapper.ParseUserMappings(userMapLines);
+
+            foreach(string line in userMapLines)
+            {
+                string[] splitLine = line.Split("=");
+                string source = splitLine[0];
+                string target = splitLine[1];
+
+                Assert.Contains(source, generatedUserMap.Keys);
```
[the remark this answers is not in the document]
Comment (Alexander-Hjelm) [reply src/WorkItemMigrator/tests/Migration.Common.Tests/JsonExtensionsTests.cs]: Fixed by b61a171

**Comment by Alexander-Hjelm** [comment]
I addressed your comments and wrote tests against null parameter values where it made sense. Can you leave your review again? :)

File: src/WorkItemMigrator/tests/Migration.Common.Tests/MigrationCommonTests.cs [hunk 16]
```diff
@@ -21,21 +20,15 @@ public void Setup()
...
         }
 
-        [Test]
-        public void When_generating_user_map_Then_map_is_correct()
+        [TestCase("a@jira.com", "a@azdo.com")]
+        [TestCase("b@jira.com", "b@azdo.com")]
+        public void When_generating_user_map_Then_map_is_correct(string source, string target)
         {
             string[] userMapLines = { "a@jira.com=a@azdo.com", "b@jira.com=b@azdo.com" };
             Dictionary<string, string> generatedUserMap = UserMapper.ParseUserMappings(userMapLines);
```
Comment (madkoo) [inline src/WorkItemMigrator/tests/Migration.Common.Tests/MigrationCommonTests.cs:29]: please wrap this into Assert.Multiple

**Review by madkoo — APPROVED** [review]
One final note, otherwise looks good :) Will approve it but please check one comment

File: src/WorkItemMigrator/tests/Migration.Common.Tests/MigrationCommonTests.cs (hunk 16 above)
Comment (Alexander-Hjelm) [reply src/WorkItemMigrator/tests/Migration.Common.Tests/MigrationCommonTests.cs:29]: Fixed by [13d30d3](https://github.com/solidify/jira-azuredevops-migrator/pull/470/commits/13d30d38b0675b475b952ea164099ce9d99eb483)
